File: wordpress_zone/tests/test_agent_requests.py
#!/usr/bin/env python3
"""
Mock PHP test runner - verifies PHP code structure and logic
Run when PHP is not available in the environment
"""

import re
import json


def test_bridge_functions_exist():
    """Verify the two functions are defined in the bridge file"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    # Check for geometry_os_send_agent_request function
    if 'function geometry_os_send_agent_request(' not in content:
        print("FAIL: geometry_os_send_agent_request function not found")
        return False
    print("PASS: geometry_os_send_agent_request function exists")
    return True


def test_geometry_os_get_task_status_exists():
    """Verify geometry_os_get_task_status function exists"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if 'function geometry_os_get_task_status(' not in content:
        print("FAIL: geometry_os_get_task_status function not found")
        return False
    print("PASS: geometry_os_get_task_status function exists")
    return True


def test_send_function_returns_array():
    """Verify send function returns array with status"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    # Check for return array pattern
    if 'return $body ??' not in content and "return ['status' => 'error']" not in content:
        print("FAIL: Function should return array")
        return False
    print("PASS: Function returns array")
    return True


def test_send_function_posts_to_visual_bridge():
    """Verify function posts to correct endpoint"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if 'http://127.0.0.1:8768/agent/request' not in content:
        print("FAIL: Function should POST to Visual Bridge agent endpoint")
        return False
    print("PASS: Function POSTs to Visual Bridge")
    return True


def test_status_function_gets_from_visual_bridge():
    """Verify status function gets from correct endpoint"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if '/agent/status/' not in content:
        print("FAIL: Function should GET from Visual Bridge status endpoint")
        return False
    print("PASS: Function GETs from Visual Bridge status endpoint")
    return True


def test_error_handling():
    """Verify error handling for connection failures"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if "is_wp_error($response)" not in content:
        print("FAIL: Function should handle WP errors")
        return False
    print("PASS: Error handling implemented")
    return True


def test_timeout_configuration():
    """Verify 5-second timeout is set"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if "'timeout' => 5" not in content:
        print("FAIL: 5-second timeout not found")
        return False
    print("PASS: Timeout configured correctly")
    return True


def test_json_content_type():
    """Verify JSON content type header"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if "'Content-Type' => 'application/json'" not in content:
        print("FAIL: JSON content type header not found")
        return False
    print("PASS: JSON content type header set")
    return True


def test_request_structure():
    """Verify request payload includes required fields"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    required_fields = ['type', 'agent_type', 'payload', 'source', 'timestamp', 'request_id']
    missing = []

    for field in required_fields:
        if f"'{field}'" not in content:
            missing.append(field)

    if missing:
        print(f"FAIL: Missing request fields: {missing}")
        return False
    print("PASS: All required request fields present")
    return True


def test_localhost_only():
    """Verify requests go to localhost only (security)"""
    with open('../wordpress/wp-content/mu-plugins/geometry_os_bridge.php', 'r') as f:
        content = f.read()

    if '127.0.0.1' not in content and 'localhost' not in content:
        print("FAIL: Should only connect to localhost")
        return False
    print("PASS: Localhost-only security constraint satisfied")
    return True


def main():
    tests = [
        test_bridge_functions_exist,
        test_geometry_os_get_task_status_exists,
        test_send_function_returns_array,
        test_send_function_posts_to_visual_bridge,
        test_status_function_gets_from_visual_bridge,
        test_error_handling,
        test_timeout_configuration,
        test_json_content_type,
        test_request_structure,
        test_localhost_only,
    ]

    print("Running Mock PHP Verification Tests")
    print("=" * 50)
    print()

    passed = 0
    failed = 0

    for test in tests:
        try:
            if test():
                passed += 1
            else:
                failed += 1
        except Exception as e:
            print(f"FAIL: {test.__name__} - Exception: {e}")
            failed += 1

    print()
    print("=" * 50)
    print(f"Results: {passed} passed, {failed} failed")

    return failed == 0


if __name__ == '__main__':
    import sys
    sys.exit(0 if main() else 1)

File: wordpress_zone/tests/test_agent_requests.php
<?php
/**
 * Test WordPress agent request handling
 *
 * Tests for geometry_os_send_agent_request() and geometry_os_get_task_status()
 * functions in the Geometry OS Bridge plugin.
 */

// Mock WordPress functions for standalone testing
if (!function_exists('wp_remote_post')) {
    function wp_remote_post($url, $args = []) {
        global $mock_wp_response;
        if (isset($mock_wp_response)) {
            return $mock_wp_response;
        }
        // Default mock response
        return [
            'response' => ['code' => 200],
            'body' => json_encode(['status' => 'queued', 'task_id' => 'test-task-123'])
        ];
    }
}

if (!function_exists('wp_remote_get')) {
    function wp_remote_get($url, $args = []) {
        global $mock_wp_response;
        if (isset($mock_wp_response)) {
            return $mock_wp_response;
        }
        return [
            'response' => ['code' => 200],
            'body' => json_encode(['status' => 'completed', 'result' => ['tags' => ['AI', 'evolution']]])
        ];
    }
}

if (!function_exists('wp_remote_retrieve_body')) {
    function wp_remote_retrieve_body($response) {
        return $response['body'] ?? '';
    }
}

if (!function_exists('wp_generate_uuid4')) {
    function wp_generate_uuid4() {
        return sprintf(
            '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            mt_rand(0, 0xffff), mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0x0fff) | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffff), mt_rand(0, 0xffff), mt_rand(0, 0xffff)
        );
    }
}

if (!function_exists('is_wp_error')) {
    function is_wp_error($thing) {
        return $thing instanceof WP_Error;
    }
}

if (!function_exists('wp_is_post_revision')) {
    function wp_is_post_revision($post) {
        return false;
    }
}

if (!function_exists('wp_get_post_tags')) {
    function wp_get_post_tags($post_id, $args = []) {
        return [];
    }
}

if (!function_exists('update_post_meta')) {
    function update_post_meta($post_id, $meta_key, $meta_value) {
        global $post_meta_storage;
        if (!isset($post_meta_storage)) {
            $post_meta_storage = [];
        }
        $post_meta_storage[$post_id][$meta_key] = $meta_value;
        return true;
    }
}

if (!function_exists('get_post_meta')) {
    function get_post_meta($post_id, $key = '', $single = false) {
        global $post_meta_storage;
        if (!isset($post_meta_storage) || !isset($post_meta_storage[$post_id])) {
            return $single ? '' : [];
        }
        if (empty($key)) {
            return $post_meta_storage[$post_id];
        }
        return $single ? ($post_meta_storage[$post_id][$key] ?? '') : [$post_meta_storage[$post_id][$key] ?? ''];
    }
}

if (!function_exists('get_posts')) {
    function get_posts($args = []) {
        return [];
    }
}

if (!function_exists('wp_next_scheduled')) {
    function wp_next_scheduled($hook) {
        return false;
    }
}

if (!function_exists('wp_schedule_event')) {
    function wp_schedule_event($timestamp, $recurrence, $hook) {
        return true;
    }
}

if (!function_exists('add_action')) {
    $GLOBALS['registered_actions'] = [];
    function add_action($hook, $callback, $priority = 10, $accepted_args = 1) {
        $GLOBALS['registered_actions'][$hook] = [
            'callback' => $callback,
            'priority' => $priority,
            'accepted_args' => $accepted_args
        ];
    }
}

if (!function_exists('add_filter')) {
    $GLOBALS['registered_filters'] = [];
    function add_filter($hook, $callback, $priority = 10, $accepted_args = 1) {
        $GLOBALS['registered_filters'][$hook] = [
            'callback' => $callback,
            'priority' => $priority,
            'accepted_args' => $accepted_args
        ];
    }
}

if (!class_exists('WP_Post')) {
    class WP_Post {
        public $ID;
        public $post_title;
        public $post_content;
        public $post_status;

        public function __construct($id = 1, $title = 'Test Post', $content = 'Test content') {
            $this->ID = $id;
            $this->post_title = $title;
            $this->post_content = $content;
            $this->post_status = 'publish';
        }
    }
}

if (!class_exists('WP_Error')) {
    class WP_Error {
        private $errors = [];
        private $error_data = [];

        public function __construct($code = '', $message = '', $data = '') {
            if (!empty($code)) {
                $this->errors[$code][] = $message;
                if (!empty($data)) {
                    $this->error_data[$code] = $data;
                }
            }
        }

        public function get_error_message($code = '') {
            if (empty($code)) {
                $code = array_key_first($this->errors);
            }
            return $this->errors[$code][0] ?? '';
        }
    }
}

// Include the bridge functions
require_once __DIR__ . '/../wordpress/wp-content/mu-plugins/geometry_os_bridge.php';

/**
 * Test class for agent request functions
 */
class TestAgentRequests {

    /**
     * Test sending content intelligence request
     */
    public static function test_send_content_intelligence_request() {
        global $mock_wp_response;
        $mock_wp_response = [
            'response' => ['code' => 200],
            'body' => json_encode(['status' => 'queued', 'task_id' => 'wp-content_intelligence-1-' . time()])
        ];

        $result = geometry_os_send_agent_request('content_intelligence', [
            'post_id' => 1,
            'action' => 'auto_tag'
        ]);

        if (!is_array($result)) {
            echo "FAIL: test_send_content_intelligence_request - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if (!array_key_exists('status', $result)) {
            echo "FAIL: test_send_content_intelligence_request - Missing 'status' key\n";
            return false;
        }
        if ($result['status'] !== 'queued') {
            echo "FAIL: test_send_content_intelligence_request - Expected 'queued', got '{$result['status']}'\n";
            return false;
        }

        echo "PASS: test_send_content_intelligence_request\n";
        return true;
    }

    /**
     * Test sending evolution publish request
     */
    public static function test_send_evolution_publish_request() {
        global $mock_wp_response;
        $mock_wp_response = [
            'response' => ['code' => 200],
            'body' => json_encode([
                'status' => 'queued',
                'task_id' => 'wp-evolution_publish-abc123-' . time()
            ])
        ];

        $result = geometry_os_send_agent_request('evolution_publish', [
            'commit_hash' => 'abc123',
            'message' => 'Test commit'
        ]);

        if (!is_array($result)) {
            echo "FAIL: test_send_evolution_publish_request - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if (!array_key_exists('task_id', $result)) {
            echo "FAIL: test_send_evolution_publish_request - Missing 'task_id' key\n";
            return false;
        }

        echo "PASS: test_send_evolution_publish_request\n";
        return true;
    }

    /**
     * Test sending plugin analysis request
     */
    public static function test_send_plugin_analysis_request() {
        global $mock_wp_response;
        $mock_wp_response = [
            'response' => ['code' => 200],
            'body' => json_encode(['status' => 'queued', 'task_id' => 'wp-plugin_analysis-' . time()])
        ];

        $result = geometry_os_send_agent_request('plugin_analysis', [
            'scan_type' => 'security'
        ]);

        if (!is_array($result)) {
            echo "FAIL: test_send_plugin_analysis_request - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if ($result['status'] !== 'queued') {
            echo "FAIL: test_send_plugin_analysis_request - Expected 'queued', got '{$result['status']}'\n";
            return false;
        }

        echo "PASS: test_send_plugin_analysis_request\n";
        return true;
    }

    /**
     * Test get task status
     */
    public static function test_get_task_status() {
        global $mock_wp_response;
        $mock_wp_response = [
            'response' => ['code' => 200],
            'body' => json_encode([
                'status' => 'completed',
                'result' => ['tags' => ['AI', 'evolution']]
            ])
        ];

        $result = geometry_os_get_task_status('wp-content_intelligence-1-1234567890');

        if (!is_array($result)) {
            echo "FAIL: test_get_task_status - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if ($result['status'] !== 'completed') {
            echo "FAIL: test_get_task_status - Expected 'completed', got '{$result['status']}'\n";
            return false;
        }

        echo "PASS: test_get_task_status\n";
        return true;
    }

    /**
     * Test error handling for connection failure
     */
    public static function test_error_handling() {
        global $mock_wp_response;
        $mock_wp_response = new WP_Error('http_request_failed', 'Connection refused');

        $result = geometry_os_send_agent_request('content_intelligence', [
            'post_id' => 1,
            'action' => 'auto_tag'
        ]);

        if (!is_array($result)) {
            echo "FAIL: test_error_handling - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if ($result['status'] !== 'error') {
            echo "FAIL: test_error_handling - Expected 'error', got '{$result['status']}'\n";
            return false;
        }

        echo "PASS: test_error_handling\n";
        return true;
    }

    /**
     * Test auto-publish on evolution commit event
     */
    public static function test_auto_publish_on_evolution_commit() {
        global $mock_wp_response, $post_meta_storage;
        $post_meta_storage = [];

        // Mock for checking duplicates (no existing posts)
        $mock_wp_response = [
            'response' => ['code' => 200],
            'body' => json_encode([
                'status' => 'queued',
                'task_id' => 'wp-evolution_publish-test-' . time()
            ])
        ];

        // Simulate evolution daemon writing commit event
        $event = [
            'type' => 'evolution_commit',
            'commit_hash' => 'abc123def',
            'message' => 'feat: add new feature'
        ];

        $result = geometry_os_handle_evolution_event($event);

        if (!is_array($result)) {
            echo "FAIL: test_auto_publish_on_evolution_commit - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if ($result['status'] !== 'published') {
            echo "FAIL: test_auto_publish_on_evolution_commit - Expected 'published', got '{$result['status']}'\n";
            return false;
        }
        if (!isset($result['task_id'])) {
            echo "FAIL: test_auto_publish_on_evolution_commit - Missing 'task_id' key\n";
            return false;
        }

        echo "PASS: test_auto_publish_on_evolution_commit\n";
        return true;
    }

    /**
     * Test evolution event handler ignores non-commit events
     */
    public static function test_evolution_event_ignores_non_commit() {
        $event = [
            'type' => 'other_event',
            'data' => 'something'
        ];

        $result = geometry_os_handle_evolution_event($event);

        if (!is_array($result)) {
            echo "FAIL: test_evolution_event_ignores_non_commit - Expected array, got " . gettype($result) . "\n";
            return false;
        }
        if ($result['status'] !== 'ignored') {
            echo "FAIL: test_evolution_event_ignores_non_commit - Expected 'ignored', got '{$result['status']}'\n";
            return false;
        }

        echo "PASS: test_evolution_event_ignores_non_commit\n";
        return true;
    }

    /**
     * Test cron schedule registration
     */
    public static function test_cron_schedule_registered() {
        global $registered_filters;

        if (!isset($registered_filters['cron_schedules'])) {
            echo "FAIL: test_cron_schedule_registered - 'cron_schedules' filter not registered\n";
            return false;
        }

        // Call the filter callback to verify it adds the five_minutes schedule
        $callback = $registered_filters['cron_schedules']['callback'];
        $schedules = $callback([]);

        if (!isset($schedules['five_minutes'])) {
            echo "FAIL: test_cron_schedule_registered - 'five_minutes' schedule not added\n";
            return false;
        }
        if ($schedules['five_minutes']['interval'] !== 300) {
            echo "FAIL: test_cron_schedule_registered - Expected interval 300, got '{$schedules['five_minutes']['interval']}'\n";
            return false;
        }

        echo "PASS: test_cron_schedule_registered\n";
        return true;
    }

    /**
     * Test auto-tag hook is registered
     */
    public static function test_auto_tag_hook_registered() {
        global $registered_actions;

        if (!isset($registered_actions['transition_post_status'])) {
            echo "FAIL: test_auto_tag_hook_registered - 'transition_post_status' action not registered\n";
            return false;
        }

        echo "PASS: test_auto_tag_hook_registered\n";
        return true;
    }

    /**
     * Test evolution polling hook is registered
     */
    public static function test_evolution_polling_hooks_registered() {
        global $registered_actions;

        if (!isset($registered_actions['init'])) {
            echo "FAIL: test_evolution_polling_hooks_registered - 'init' action not registered\n";
            return false;
        }
        if (!isset($registered_actions['geometry_os_poll_evolution'])) {
            echo "FAIL: test_evolution_polling_hooks_registered - 'geometry_os_poll_evolution' action not registered\n";
            return false;
        }

        echo "PASS: test_evolution_polling_hooks_registered\n";
        return true;
    }

    /**
     * Run all tests
     */
    public static function run_all() {
        $tests = [
            'test_send_content_intelligence_request',
            'test_send_evolution_publish_request',
            'test_send_plugin_analysis_request',
            'test_get_task_status',
            'test_error_handling',
            'test_auto_publish_on_evolution_commit',
            'test_evolution_event_ignores_non_commit',
            'test_cron_schedule_registered',
            'test_auto_tag_hook_registered',
            'test_evolution_polling_hooks_registered'
        ];

        $passed = 0;
        $failed = 0;

        echo "Running Agent Request Tests\n";
        echo str_repeat("=", 50) . "\n\n";

        foreach ($tests as $test) {
            try {
                if (self::$test()) {
                    $passed++;
                } else {
                    $failed++;
                }
            } catch (Exception $e) {
                echo "FAIL: $test - Exception: " . $e->getMessage() . "\n";
                $failed++;
            }
        }

        echo "\n" . str_repeat("=", 50) . "\n";
        echo "Results: $passed passed, $failed failed\n";

        return $failed === 0;
    }
}

// Run tests if executed directly
if (php_sapi_name() === 'cli' && basename(__FILE__) === basename($_SERVER['SCRIPT_NAME'] ?? '')) {
    $success = TestAgentRequests::run_all();
    exit($success ? 0 : 1);
}
